src/storage/v2/vertex_info_helpers.hpp

```diff
@@ -0,0 +1,215 @@
+// Copyright 2023 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+#pragma once
+
+#ifndef MEMGRAPH__STORAGE__VERTEX_INFO_HELPERS_HPP
+#define MEMGRAPH__STORAGE__VERTEX_INFO_HELPERS_HPP
+
+#include "storage/v2/delta.hpp"
+#include "storage/v2/edge_direction.hpp"
+#include "storage/v2/vertex_info_cache.hpp"
+#include "utils/variant_helpers.hpp"
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+namespace memgraph::storage {
+
+template <Delta::Action>
+struct DeltaAction_tag {};
+
+template <Delta::Action A, typename Method>
+struct ActionMethodImpl : Method {
+  // Uses tag dispatch to ensure method is only called for the correct action
+  void operator()(DeltaAction_tag<A> /*unused*/, Delta const &delta) { Method::operator()(delta); }
+};
+
+template <Delta::Action A, typename Method>
+inline auto ActionMethod(Method &&func) {
+  return ActionMethodImpl<A, Method>{std::forward<Method>(func)};
+}
+
+/// Converts runtime Delta::Action into compile time tag, this allows us to dispatch to the correct overload
+template <typename Func>
+inline void DeltaDispatch(Delta const &delta, Func &&func) {
+  // clang-format off
+#define dispatch(E) case E: return func(DeltaAction_tag<E>{}, delta); // NOLINT
+  // clang-format on
+  switch (delta.action) {
+    using enum Delta::Action;
+    dispatch(DELETE_DESERIALIZED_OBJECT);
+    dispatch(DELETE_OBJECT);
+    dispatch(RECREATE_OBJECT);
+    dispatch(SET_PROPERTY);
+    dispatch(ADD_LABEL);
+    dispatch(REMOVE_LABEL);
+    dispatch(ADD_IN_EDGE);
+    dispatch(ADD_OUT_EDGE);
+    dispatch(REMOVE_IN_EDGE);
+    dispatch(REMOVE_OUT_EDGE);
+  }
+#undef dispatch
+}
+
+inline auto Exists_ActionMethod(bool &exists) {
+  using enum Delta::Action;
+  // clang-format off
+  return utils::Overloaded{
+      ActionMethod<DELETE_DESERIALIZED_OBJECT>([&](Delta const & /*unused*/) { exists = false; }),
+      ActionMethod<DELETE_OBJECT>([&](Delta const & /*unused*/) { exists = false; })
+  };
+  // clang-format on
+}
+
+inline auto Deleted_ActionMethod(bool &deleted) {
+  using enum Delta::Action;
+  return ActionMethod<RECREATE_OBJECT>([&](Delta const & /*unused*/) { deleted = false; });
+}
+
+inline auto HasLabel_ActionMethod(bool &has_label, LabelId label) {
+  using enum Delta::Action;
+  // clang-format off
+  return utils::Overloaded{
+      ActionMethod<REMOVE_LABEL>([&, label](Delta const &delta) {
+        if (delta.label == label) {
+          MG_ASSERT(has_label, "Invalid database state!");
```

Comment:
These MG_ASSERTS are concerning. `has_label` is an input, we can't control it here.

Reply:
Its an in + out reference. This is logic that remains from pre refactor.